File: src/main/java/uk/co/endofhome/javoice/invoice/Invoice.java
package uk.co.endofhome.javoice.invoice;

import com.googlecode.totallylazy.Sequence;
import uk.co.endofhome.javoice.customer.Customer;

import java.time.LocalDate;

public class Invoice {
    public final String number;
    public final LocalDate date;
    public final Customer customer;
    public final String orderNumber;
    public final Sequence<ItemLine> itemLines;

    public static final int ITEM_LINES_START_AT = 17;
    public static final int MAX_ITEM_LINES = 17;

    public Invoice(String invoiceNumber, LocalDate date, Customer customer, String orderNumber, Sequence<ItemLine> itemLines) {
        this.number = invoiceNumber;
        this.date = date;
        this.customer = customer;
        this.orderNumber = orderNumber;
        this.itemLines = itemLines;
    }

    public double nettValue() {
        if (itemLines.isEmpty()) {
            return 0;
        }
        return itemLines.map((item -> (item.quantity.getOrElse(0.0) * item.unitPrice.getOrElse(0.0))))
                .reduce((x, y) -> ((double) x) + y);
    }

    public String customerName() {
        return customer.name;
    }
}

File: src/main/java/uk/co/endofhome/javoice/invoice/Invoice.kt
package uk.co.endofhome.javoice.invoice

import uk.co.endofhome.javoice.customer.Customer
import java.time.LocalDate

class Invoice(val number: String, val date: LocalDate, val customer: Customer, val orderNumber: String, val itemLines: List<ItemLine>) {

    fun nettValue(): Double {
        return when (itemLines.isNotEmpty()) {
            true -> itemLines
                .map { ((it.quantity ?: 0.0) * (it.unitPrice ?: 0.0)) }
                .reduce { x, y -> x + y }
            false -> 0.0
        }
    }

    fun customerName(): String {
        return customer.name
    }

    companion object {
        val ITEM_LINES_START_AT = 17
        val MAX_ITEM_LINES = 17
    }
}
